### ISSUE TITLE
error: 'mps.add' op requires the same element type for all operands and results

### ISSUE THREAD

**Issue by randyoo** [gregcmartin/HunyuanVideo_MLX#1]
On a M4 Max running 15.1.1, when running sample_video_mps.py:

```
2024-12-16 23:07:29.108 | INFO     | hyvideo.vae:load_vae:29 - Loading 3D VAE model (884-16c-hy) from: ./ckpts/hunyuan-video-t2v-720p/vae
/Users/admin/Desktop/hunyuanvideo/HunyuanVideo_MLX/hyvideo/vae/__init__.py:39: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  ckpt = torch.load(vae_ckpt, map_location=vae.device)
2024-12-16 23:07:29.630 | INFO     | hyvideo.vae:load_vae:55 - VAE to dtype: torch.float32
2024-12-16 23:07:30.220 | INFO     | hyvideo.text_encoder:load_text_encoder:28 - Loading text encoder model (llm) from: ./ckpts/text_encoder
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 17.03it/s]
2024-12-16 23:07:30.771 | INFO     | hyvideo.text_encoder:load_text_encoder:50 - Text encoder to dtype: torch.float32
2024-12-16 23:07:41.566 | INFO     | hyvideo.text_encoder:load_tokenizer:64 - Loading tokenizer (llm) from: ./ckpts/text_encoder
2024-12-16 23:07:41.746 | INFO     | hyvideo.text_encoder:load_text_encoder:28 - Loading text encoder model (clipL) from: ./ckpts/text_encoder_2
2024-12-16 23:07:41.971 | INFO     | hyvideo.text_encoder:load_text_encoder:50 - Text encoder to dtype: torch.float16
2024-12-16 23:07:42.319 | INFO     | hyvideo.text_encoder:load_tokenizer:64 - Loading tokenizer (clipL) from: ./ckpts/text_encoder_2
2024-12-16 23:07:42.327 | INFO     | hyvideo.inference:predict:577 - Input (height, width, video_length) = (544, 960, 129)
2024-12-16 23:07:42.343 | DEBUG    | hyvideo.inference:predict:637 - 
                        height: 544
                         width: 960
                  video_length: 129
                        prompt: ['a cat is running, realistic.']
                    neg_prompt: ['Aerial view, aerial view, overexposed, low quality, deformation, a poor composition, bad hands, bad teeth, bad eyes, bad limbs, distortion']
                          seed: None
                   infer_steps: 40
         num_videos_per_prompt: 1
                guidance_scale: 7.0
                      n_tokens: 67320
                    flow_shift: 7.0
       embedded_guidance_scale: 6.0
  0%|                                                                                                                                                                            | 0/40 [00:00<?, ?it/s](mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf16>, tensor<3072xf32>) -> tensor<*xf32>
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf16>, tensor<3072xf32>) -> tensor<*xf32>
/AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:988: failed assertion `original module failed verification'
```

Seems the conflicting argument types are fp16 and fp32.

Doubt it's related, but in case it is, when I run check_system.py, I get:

```
=== Model Weights ===

⚠️  Missing model weights:
    - vae/884-16c-hy.pt
    - text_encoder/llm.pt
```

That happens despite successfully running 'python download_weights.py' with no errors.

**Comment by alexayers** [comment]
Also on M4 Max and hitting same issue.

**Comment by gregcmartin** [comment]
The fp16 bug fp32 was fixed last night make sure to pull the latest code down and try again

**Comment by alexayers** [comment]
I only found out about repo an hour ago. :-) I'm on main, should I be on a different branch.

**Comment by alexayers** [comment]
Looks like my output is slightly different:

`2024-12-17 10:54:15.100 | INFO     | hyvideo.text_encoder:load_text_encoder:50 - Text encoder to dtype: torch.float16
2024-12-17 10:54:16.772 | INFO     | hyvideo.text_encoder:load_tokenizer:64 - Loading tokenizer (clipL) from: ./ckpts/text_encoder_2
2024-12-17 10:54:16.779 | INFO     | hyvideo.inference:predict:577 - Input (height, width, video_length) = (544, 960, 129)
2024-12-17 10:54:16.807 | DEBUG    | hyvideo.inference:predict:637 - 
                        height: 544
                         width: 960
                  video_length: 129
                        prompt: ['a cat is running, realistic.']
                    neg_prompt: ['Aerial view, aerial view, overexposed, low quality, deformation, a poor composition, bad hands, bad teeth, bad eyes, bad limbs, distortion']
                          seed: None
                   infer_steps: 30
         num_videos_per_prompt: 1
                guidance_scale: 7.0
                      n_tokens: 67320
                    flow_shift: 7.0
       embedded_guidance_scale: 6.0
  0%|                                                                                                                                                                            | 0/30 [00:00<?, ?it/s]/Users/alex/github/HunyuanVideo_MLX/venv/lib/python3.10/site-packages/torch/amp/autocast_mode.py:266: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn(
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf32>, tensor<3072xf16>) -> tensor<*xf32>
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf32>, tensor<3072xf16>) -> tensor<*xf32>
/AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:988: failed assertion `original module failed verification'
zsh: abort      python sample_video_mps.py --video-size 544 960 --video-length 129  30     
(venv) (base) alex@Alexs-MacBook-Pro-2 HunyuanVideo_MLX % /opt/homebrew/Cellar/python@3.10/3.10.16/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '`

**Comment by randyoo** [comment]
> The fp16 bug fp32 was fixed last night make sure to pull the latest code down and try again

Started from scratch, install went smoothly this time, but still running into this mps.add fp16/fp32 issue.

```
2024-12-17 10:32:03.323 | INFO     | hyvideo.inference:from_pretrained:153 - Got text-to-video model root path: ckpts
2024-12-17 10:32:03.323 | INFO     | hyvideo.inference:from_pretrained:188 - Building model...
2024-12-17 10:32:04.121 | INFO     | hyvideo.inference:load_state_dict:337 - Loading torch model ckpts/hunyuan-video-t2v-720p/transformers/mp_rank_00_model_states.pt...
/Users/admin/Downloads/HunyuanVideo_MLX/hyvideo/inference.py:338: FutureWarning: (snip)
2024-12-17 10:32:09.805 | INFO     | hyvideo.vae:load_vae:29 - Loading 3D VAE model (884-16c-hy) from: ./ckpts/hunyuan-video-t2v-720p/vae
/Users/admin/Downloads/HunyuanVideo_MLX/hyvideo/vae/__init__.py:39: FutureWarning: (snip)
2024-12-17 10:32:10.364 | INFO     | hyvideo.vae:load_vae:55 - VAE to dtype: torch.float16
2024-12-17 10:32:10.950 | INFO     | hyvideo.text_encoder:load_text_encoder:28 - Loading text encoder model (llm) from: ./ckpts/text_encoder
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 11.77it/s]
2024-12-17 10:32:13.456 | INFO     | hyvideo.text_encoder:load_text_encoder:50 - Text encoder to dtype: torch.float16
2024-12-17 10:32:15.049 | INFO     | hyvideo.text_encoder:load_tokenizer:64 - Loading tokenizer (llm) from: ./ckpts/text_encoder
2024-12-17 10:32:15.209 | INFO     | hyvideo.text_encoder:load_text_encoder:28 - Loading text encoder model (clipL) from: ./ckpts/text_encoder_2
2024-12-17 10:32:15.298 | INFO     | hyvideo.text_encoder:load_text_encoder:50 - Text encoder to dtype: torch.float16
2024-12-17 10:32:15.818 | INFO     | hyvideo.text_encoder:load_tokenizer:64 - Loading tokenizer (clipL) from: ./ckpts/text_encoder_2
2024-12-17 10:32:15.825 | INFO     | hyvideo.inference:predict:577 - Input (height, width, video_length) = (720, 1280, 129)
2024-12-17 10:32:15.840 | DEBUG    | hyvideo.inference:predict:637 - 
                        height: 720
                         width: 1280
                  video_length: 129
                        prompt: ['a chicken walks around']
                    neg_prompt: ['Aerial view, aerial view, overexposed, low quality, deformation, a poor composition, bad hands, bad teeth, bad eyes, bad limbs, distortion']
                          seed: None
                   infer_steps: 30
         num_videos_per_prompt: 1
                guidance_scale: 7.0
                      n_tokens: 118800
                    flow_shift: 7.0
       embedded_guidance_scale: 6.0
  0%|                                                                                                                          | 0/30 [00:00<?, ?it/s]/Users/admin/Downloads/HunyuanVideo_MLX/venv/lib/python3.10/site-packages/torch/amp/autocast_mode.py:266: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn(
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf32>, tensor<3072xf16>) -> tensor<*xf32>
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: error: 'mps.add' op requires the same element type for all operands and results
(mpsFileLoc): /AppleInternal/Library/BuildRoots/c4b332c0-77a5-11ef-9847-daac7d5d70b1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:233:0: note: see current operation: %7 = "mps.add"(%5, %arg2) : (tensor<2x3072xf32>, tensor<3072xf16>) -> tensor<*xf32>
```

**Comment by gregcmartin** [comment]
made updates generates fine now but using CPU only, working on MLX optimization for GPU offload so its not painfully slow.  Also fixed the issue with the FP mismatch and the check weights error
